content/blog/2019-09-25-interpreter-vector-support/index.md

```diff
@@ -0,0 +1,134 @@
+
+Memory access requires both loads and store operations. To emulate arrays, we can loop over multiple addresses and perform a load and/or store at each location. Load and store instructions in Bril are implemented similarly to their assembly counterparts. A load takes in a memory address (an `int` register in Bril) and writes to a destination register. A store is an effect operation and does not write to a register. It uses two source registers: a register containing a value and a register containing an address. The Bril syntax is given below.
+
+```C
+// load value from address 2
+addr: int = const 2;
+data: int = lw addr;
+
+// store value to address 2
+sw data addr;
+```
+
+### Interpreted Vector Instructions
+
+Vector instructions were added to the Bril interpreter. We specifically implement fixed-sized vector instructions of length four (akin to native Intel `__m128` SSE instructions). Typescript and Javascript (Typescript always compiles to Javascript) do not have support for vector instrinsics in the current standard. Thus, we implement the Bril vector instructions as a loop over four values. Additionally, we add vector registers in Bril which must be used in the vector instructions. We target a vector-vector add (vvadd) program, so we include interpreter support for `vadd`, `vload`, and `vstore` instructions. The `vload` and `vstore` instructions communicate data between vector registers and the interpreter stack. The `vadd` instructions adds two vector registers and writes to a destination vector register. An example vvadd program is shown below.
+
+```C
+// locations of memory (arrays of 4 elements)
+a: int = const 0;
+b: int = const 4;
+c: int = const 8;
+
+// load into src vector registers
+va: vector = vload a;
+vb: vector = vload b;
+
+// do the vector add
+vc: vector = vadd va vb;
+
+// store from vector register to memory
+vstore vc c;
+```
+
+### Intrinsic Vector Support
+
+It's awkward that the interpreter supports vector instructions, but doesn't actually use a native vector assembly instruction to perform the computation. We expect that the performance of the interpreted vector instructions will be poor. To explore this hypothesis, we create a version of an interpreted vector add instruction using three methods: 1) Typescript, 2) serial c++, and 3) vectorized c++. We run each test for 10,000 iterations and average the execution time over five runs. We assumed 10,000 iterations was enough time for the Typescript/Javascript JIT to warmup. The average time per loop iteration for each implementation is given below.
```

Comment:
```suggestion
It's awkward that the interpreter supports vector instructions but doesn't actually use a native vector assembly instruction to perform the computation. We expect that the performance of the interpreted vector instructions will be poor. To explore this hypothesis, we create a version of an interpreted vector add instruction using three methods: 1) TypeScript, 2) serial C++, and 3) vectorized C++. We run each test for 10,000 iterations and average the execution time over five runs. We assumed 10,000 iterations was enough time for the TypeScript/JavaScript JIT to warmup. The average time per loop iteration for each implementation is given below.
```

Dependent clauses => no comma. Also C++ is capitalized.

Reply:
To be specific about the JIT, you're using [Node](https://nodejs.org/en/), which uses the [Google's V8 JIT](https://v8.dev) to execute JS.